sdk/core/core-rest-pipeline/src/pipeline.ts

```diff
@@ -155,12 +158,15 @@ class HttpPipeline implements Pipeline {
         return true;
       }
     });
-    this._orderedPolicies = undefined;
+    this.damage();
 
     return removedPolicies;
   }
 
   public sendRequest(httpClient: HttpClient, request: PipelineRequest): Promise<PipelineResponse> {
+    const _cached = this._cache.get(httpClient);
```

Comment:
Totally up to you but what I like to do with cached data is to _always_ return the cached data. So I do something like:

```ts
if (!this._cache.has(aThing)) { 
   const theThing = ... expensive operation
   this._cache.set(aThing, theThing)
}
return this._cache.get(aThing); 
```

But it's purely a stylistic choice so feel free to ignore (or convince me not to that anymore either 😄)

Reply:
The reason I didn't do this is because it seems to require two weakmap lookups instead of one per call. One for `has` and one for `get`. That said, there doesn't empirically seem to be much difference at all. I don't know enough about the implementation of `WeakMap` to guess about why there isn't much difference, but I'm open to this idea.

Reply:
If the goal is to avoid multiple returns (which there are some stylistic arguments for) perhaps something like

```ts
let pipeline = this._cache.get(httpClient);
if (!pipeline) {
   // calculate new function
   this._cache.set(httpClient, pipeline);
}
return pipeline(request);
```

would suffice?

Reply:
I like that. There is a TC39 proposal that would make something like the following possible, which would be useful in cases like these:

```typescript
const pipeline = this.cache.get(httpClient) ?? do {
   // calculate new function
   this._cache.set(httpClient, nextPipeline);
   // `do` blocks evaluate to the last ExpressionStatement executed in the block
   nextPipeline;
};

return pipeline(request);
```

Reply:
> The reason I didn't do this is because it seems to require two weakmap lookups instead of one per call. One for `has` and one for `get`. That said, there doesn't empirically seem to be much difference at all. I don't know enough about the implementation of `WeakMap` to guess about why there isn't much difference, but I'm open to this idea.

I usually don't worry about map lookups as they are so cheap - you can probably do a million of them for the cost of one network hop. But I know what you mean. And feel free to correct my assumptions if they're wrong, I obviously didn't benchmark before making this statement 😆 

> I like that. There is a TC39 proposal that would make something like the following possible, which would be useful in cases like these:
> 
> ```ts
> const pipeline = this.cache.get(httpClient) ?? do {
>    // calculate new function
>    this._cache.set(httpClient, nextPipeline);
>    // `do` blocks evaluate to the last ExpressionStatement executed in the block
>    nextPipeline;
> };
> 
> return pipeline(request);
> ```

Reminds me of C# code I used to write in 2014 (I even did this SO answer when I still cared about SO) https://stackoverflow.com/questions/278703/unique-ways-to-use-the-null-coalescing-operator/24309582#24309582